From f3a266fd093112079f59ca7bd7c400dd97d0ace0 Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Thu, 26 Dec 2019 12:04:36 +0800
Subject: [PATCH] DRVLIB_SW-3240: Add API in RCU driver to report reset cause

Used single global static variable to store and retrieve the reset cause
---
 drivers/reset/reset-intel-syscon.c | 26 +++++++++++++++++++-------
 1 file changed, 19 insertions(+), 7 deletions(-)

diff --git a/drivers/reset/reset-intel-syscon.c b/drivers/reset/reset-intel-syscon.c
index 9176400157e4..f028d1240796 100644
--- a/drivers/reset/reset-intel-syscon.c
+++ b/drivers/reset/reset-intel-syscon.c
@@ -35,14 +35,19 @@ struct intel_reset_data {
 	u32 *rst_ids;
 	u32 rst_num;
 	struct dentry *debugfs;
-	char reset_cause[STR_MAX_LEN];
 };
 
-static char rst_cause[STR_MAX_LEN] = "";
+#if IS_ENABLED(CONFIG_DEBUG_FS)
+static char rst_cause[STR_MAX_LEN];
+#endif /* CONFIG_DEBUG_FS */
 
+#if IS_ENABLED(CONFIG_DEBUG_FS)
 static int __init get_reset_cause(char *line)
 {
-	strlcpy(rst_cause, line, STR_MAX_LEN - 1);
+	size_t len = strlen(line);
+
+	memcpy(rst_cause, line,
+	       ((len < (STR_MAX_LEN - 1)) ? len : (STR_MAX_LEN - 1)));
 	return 1;
 }
 
@@ -50,9 +55,10 @@ __setup("rst_cause=", get_reset_cause);
 
 static int rst_cause_read(struct seq_file *s, void *v)
 {
-	struct intel_reset_data *rst_data = s->private;
+	size_t len = strlen(rst_cause);
+
+	seq_printf(s, "reset_cause: %s\n", ((len > 0) ? rst_cause : "None"));
 
-	seq_printf(s, "%s\n", rst_data->reset_cause);
 	return 0;
 }
 
@@ -79,6 +85,12 @@ static int reset_debugfs_init(struct intel_reset_data *priv)
 	dev_err(priv->dev, "reset debugfs create fail!\n");
 	return ret;
 }
+#else
+static int reset_debugfs_init(struct intel_reset_data *priv)
+{
+	return 0;
+}
+#endif /* CONFIG_DEBUG_FS */
 
 static u32 intel_stat_reg_off(struct intel_reset_data *data, u32 req_off)
 {
@@ -266,9 +278,9 @@ static int intel_reset_probe(struct platform_device *pdev)
 	if (err)
 		dev_warn(dev, "Failed to register restart handler\n");
 
-	memcpy(data->reset_cause, rst_cause, STR_MAX_LEN);
-
 	err = reset_debugfs_init(data);
+	if (err)
+		dev_warn(dev, "Failed to register debugfs\n");
 
 	return err;
 }
